lib/Common/Memory/MarkContext.inl

```diff
@@ -205,42 +205,45 @@ void MarkContext::ProcessMark()
     while (!markStack.IsEmpty() || !preciseStack.IsEmpty())
 #endif
     {
+        if (!markStack.IsEmpty())
```

Comment:
Well, this check was not necessary before?
And looking at the PageStack<T>::Pop, it supports poping when the markstack is empty.

The only case it doesn't support is that the PageStack isn't initialized (and thus currentChunk is nullptr) which shouldn't happen here, right?

Reply:
Did you reach this method before if you didn't allocate anything?  If I edit GCStress to only allocate RecyclerVisitedHost objects it will assert that currentChunk == null when trying to Pop here in ProcessMark.

Reply:
I believe Curtis' point is that if we have anything in either stack, we'd expect them both to be initialized. I do have the repro locally and am working through how we get into this state. I'll reply back with what I find.

Reply:
This is due to the way we perform stack splitting for parallel mark. Now that we split both of the stacks, we can get into a situation where the target stacks of the Split operation are never initialized since there is nothing in the original markStack (or preciseStack). Even though those stacks are initialized, the target stacks will not be if there is nothing to transfer. If you think there's a better way to guard against this, please let me know, but this check seems valid to me. The relevant code is as follows:

    uint MarkContext::Split(uint targetCount, __in_ecount(targetCount) MarkContext ** targetContexts)
    {
        ...
        // Return the max count of the two splits - since the stacks have more or less unrelated sizes, they
        // could yield different number of splits, but the caller wants to know the max parallelism it
        // should use on the results of the split.
        const uint markStackSplitCount = this->markStack.Split(targetCount, targetMarkStacks);
#ifdef RECYCLER_VISITED_HOST
        const uint preciseStackSplitCount = this->preciseStack.Split(targetCount, targetPreciseStacks);
        return max(markStackSplitCount, preciseStackSplitCount);
#else
        return markStackSplitCount;
#endif

Reply:
Sounds good.  Please put in a comment.